[STAGE-6] SAP BTP Multitenancy Advanced

polinatonko opened the pull request:
Complete below steps:
1. Implement request filter to set current tenant info (tenantId, subdomain) from JWT into request context
2. Implement logic for new database schema creation/deletion on tenant subscription/unsubscription locally
3. Implement logic for new HANA schema creation/deletion via Service Manager API on tenant subscription/unsubscription in cloud
4. Adapt Liquibase schema migration to multitenant mode
5. Configure connection to required schema using Hibernate
6. Implement data source with connection pools per tenant
7. Adjust logic with Destination service integration to multitenant mode (i.e. try to find destination by name on subscriber subaccount level firstly, if not presented – on provider subaccount level)
8. Implement endpoint /dependencies that returns Destination service xsappname during subscription
9. Create Destination with SMTP server credentials on the Subscriber subaccount level with the same name as at Provider subaccount level

polinatonko committed:
feat: add request filter to extract subdomain value from jwt; remove xsuaa authentication from actuator endpoints

polinatonko committed:
feat: implement logic for new database schema creation/deletion on tenant subscription/unsubscription locally

polinatonko committed:
feat: implement logic for new HANA schema creation/deletion via Service Manager API on tenant subscription/unsubscription in cloud

polinatonko committed:
fix: update handling authentication errors for basic authentication

polinatonko committed:
fix: update cache for access tokens

polinatonko committed:
fix: identifying tenant locally and in cloud

polinatonko committed:
feat: adapt liquibase schema migration to multitenant mode

polinatonko committed:
refactor: refactor btp classes

polinatonko committed:
feat: configure connection to required schema using hibernate locally

polinatonko committed:
chore: add database_id parameter when creating service instance

polinatonko committed:
refactor: add hibernate properties customizer class

polinatonko committed:
feat: add saving subdomain into tenant context in cloud

polinatonko committed:
feat: implement multitenancy on the database level

polinatonko committed:
feat: adjust logic with Destination service integration to multitenant mode

polinatonko committed:
refactor: refactor multitenancy

polinatonko committed:
docs: add javadoc to classes in the servicemanager package

polinatonko committed:
docs: add javadoc to classes from the multitenancy package

polinatonko committed:
chore: remove unused rest client method

polinatonko committed:
docs: add javadoc to the xsuaa btp package

polinatonko committed:
Merge branch 'final-stage-5' into stage-6

polinatonko committed:
fix: edit service manager uri builder

polinatonko committed:
docs: update postman collection

polinatonko committed:
refactor: usage of optional class and lombok annotations

polinatonko committed:
refactor: data source configuration

polinatonko committed:
refactor: database schema management locally

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/db/service/TenantManagementService.java:
```diff
@@ -0,0 +1,58 @@
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class TenantManagementService {
```
class should be called `DatabaseManagementService` or `TenantDatabaseManagementService` because currently name is too abstract, we need to narrow down responsibility of this service

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/db/service/TenantManagementService.java:
```diff
@@ -0,0 +1,58 @@
+     * @param tenantId  the tenant identifier
+     * @param subdomain the tenant's subdomain
+     */
+    public void create(String tenantId, String subdomain) {
```
Right now purpose of this method is not obvious (we need to rename class) and then rename method to `createSchema`

SN4NTR commented on src/main/java/org/example/learningsystem/btp/servicemanager/common/builder/ServiceManagerURIBuilder.java:
```diff
@@ -0,0 +1,88 @@
+@Component
+@Profile("cloud")
+@RequiredArgsConstructor
+public class ServiceManagerURIBuilder {
```
Based on Java naming convention, name should be `ServiceManagerUriBuilder`

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/db/schema/CloudTenantSchemaService.java:
```diff
@@ -0,0 +1,51 @@
+        var schemaName = buildInstanceName(tenantInfo.tenantId());
+        var bindingName = buildBindingName(tenantInfo.tenantId());
+
+        var instance = serviceInstanceManager.createByOfferingAndPlanName(schemaName, "hana", "schema");
```
better to name `serviceInstance`

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/db/schema/CloudTenantSchemaService.java:
```diff
@@ -0,0 +1,51 @@
+        log.info("Created service instance {}", instance);
+
+        var instanceId = instance.id();
+        var binding = serviceBindingManager.create(bindingName, instanceId, tenantInfo);
```
`serviceBinding`

SN4NTR commented on src/main/resources/application-cloud.yaml:
```diff
@@ -24,6 +24,9 @@ mail:
   start-tls-enable: ${vcap.services.lms-user-service.credentials.startTls}
 
+multitenancy:
+  default-tenant: 724669da-9da4-479b-a040-cbee69ac27d1
```
it should be moved to environment variables

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/context/TenantContext.java:
```diff
@@ -0,0 +1,42 @@
+ * Thread-local storage for current tenant information.
+ */
+@NoArgsConstructor(access = PRIVATE)
+public class TenantContext {
```
It's better to name RequestContext, to make it more abstract. And here we will be able to put any data related to request (e.g. current tenant, ip address of caller and so on)

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/filter/LocalTenantResolver.java:
```diff
@@ -0,0 +1,26 @@
+ */
+@Component
+@Profile("!cloud")
+public class LocalTenantResolver implements TenantResolver {
```
resolver should be in another package called `resolver`

SN4NTR commented on src/main/java/org/example/learningsystem/multitenancy/filter/TenantIdentifierFilter.java:
```diff
@@ -0,0 +1,54 @@
+@Component
+@Slf4j
+@RequiredArgsConstructor
+public class TenantIdentifierFilter extends OncePerRequestFilter {
```
Better to name `RequestContextFilter`

SN4NTR replied:
It's also required to define order of this filter when it should be executed. It should be after `apiSecurityFilterChain` (to be able to work verified JWT). Also Spring provides `FilterRegistrationBean` to programmatically register your filter

SN4NTR commented on src/main/java/org/example/learningsystem/btp/destinationservice/service/CloudDestinationServiceImpl.java:
```diff
@@ -29,17 +33,29 @@ public class CloudDestinationServiceImpl implements DestinationService {
+            } catch (NotFound e) {
+                log.info("Destination {} wasn't found for current tenant", name);
+            }
+        }
```
it's better to split this into 2 methods, where you will try to find destination in tenant subaccount and only then on provider

polinatonko committed:
refactor: refactor stage-6

polinatonko committed:
fix: several small fixes

Update the request context filter configuration. Add closure of
tenants' data sources on application shutdown. Add exception handling
to the email service. Minor refactoring.
